src/esp32/webhook.rs

```
@@ -0,0 +1,128 @@
+            crt_bundle_attach: Some(esp_idf_sys::esp_crt_bundle_attach),
+            ..Default::default()
+        })
+        .map_err(|e| WebhookError::Other(e.to_string()))?,
+    );
+
+    webhook.send_request(&mut client)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Error)]
+pub enum WebhookError {
+    #[error("`webhook` not present in board attributes")]
+    NoWebhook,
+    #[error("`webhook-secret` not present in board attributes")]
+    NoWebhookSecret,
+    #[error("error in config: {0}")]
+    ConfigError(String),
+    #[error("failed to send request")]
+    RequestError,
+    #[error("did not receive successful response")]
+    ResponseError,
+    #[error("{0}")]
+    Other(String),
+}
+
+#[derive(Debug, Default)]
+pub struct Webhook {
+    /// Fully qualified domain name (or URL) of the robot
+    fqdn: String,
+    /// Location Secret used by the webhook's SDK script to connect
+    secret: String,
+    /// Endpoint that will be sent a GET request with credential information
+    endpoint: String,
+}
+
+impl Webhook {
+    pub fn from_robot_config(config: RobotConfig) -> Result<Self, WebhookError> {
+        let components = config.components;
+        let cloud = config.cloud.as_ref().ok_or_else(|| {
+            WebhookError::ConfigError("robot config does not have cloud config".to_string())
+        })?;
+        let fqdn = cloud.fqdn.clone();
+        let board_cfg: DynamicComponentConfig = components
+            .iter()
+            .find(|x| x.r#type == "board")
+            .ok_or_else(|| {
+                WebhookError::ConfigError("board component not found in robot config".to_string())
+            })?
+            .try_into()
+            .map_err(|_| {
+                WebhookError::ConfigError(
+                    "could not convert board config to DynamicComponentConfig".to_string(),
+                )
+            })?;
+
+        let endpoint = board_cfg
+            .get_attribute::<String>("webhook")
```

I think the logic should be to have a silent failure if the key is absent and return an error if we couldn't convert into a String

---

also check if the string is a valid URL